lib/std/crypto/sha3.zig

```diff
@@ -1,181 +1,144 @@
 
-    for (&s, 0..) |*r, i| {
-        r.* = mem.readIntLittle(u64, d[8 * i ..][0..8]);
-    }
+        st: State = .{},
+        buf: [State.rate]u8 = undefined,
+        offset: usize = 0,
 
-    for (RC[0..no_rounds]) |round| {
-        // theta
-        comptime var x: usize = 0;
-        inline while (x < 5) : (x += 1) {
-            c[x] = s[x] ^ s[x + 5] ^ s[x + 10] ^ s[x + 15] ^ s[x + 20];
+        /// The recommended output length, in bytes.
+        pub const digest_length = security_level / 2;
+        /// The block length, or rate, in bytes.
+        pub const block_length = State.rate;
+        /// Keccak does not have any options.
+        pub const Options = struct {};
+
+        /// Initialize a SHAKE extensible hash function.
+        pub fn init(options: Options) Self {
+            _ = options;
+            return Self{};
         }
-        x = 0;
-        inline while (x < 5) : (x += 1) {
-            t[0] = c[M5[x + 4]] ^ math.rotl(u64, c[M5[x + 1]], @as(usize, 1));
-            comptime var y: usize = 0;
-            inline while (y < 5) : (y += 1) {
-                s[x + y * 5] ^= t[0];
-            }
+
+        /// Hash a slice of bytes.
+        /// `out` can be any length.
+        pub fn hash(bytes: []const u8, out: []u8, options: Options) void {
+            var st = Self.init(options);
+            st.update(bytes);
+            st.squeeze(out);
         }
 
-        // rho+pi
-        t[0] = s[1];
-        x = 0;
-        inline while (x < 24) : (x += 1) {
-            c[0] = s[PIL[x]];
-            s[PIL[x]] = math.rotl(u64, t[0], ROTC[x]);
-            t[0] = c[0];
+        /// Absorb a slice of bytes into the state.
+        pub fn update(self: *Self, bytes: []const u8) void {
+            self.st.absorb(bytes);
         }
 
-        // chi
-        comptime var y: usize = 0;
-        inline while (y < 5) : (y += 1) {
-            x = 0;
-            inline while (x < 5) : (x += 1) {
-                c[x] = s[x + y * 5];
+        /// Squeeze a slice of bytes from the state.
+        /// `out` can be any length, and the function can be called multiple times.
+        pub fn squeeze(self: *Self, out_: []u8) void {
+            var out = out_;
+            if (self.offset > 0) {
+                const left = self.buf.len - self.offset;
+                if (left > 0) {
+                    const n = math.min(left, out.len);
+                    mem.copy(u8, out[0..n], self.buf[self.offset..][0..n]);
+                    out = out[n..];
+                    self.offset += n;
+                    if (out.len == 0) {
+                        return;
+                    }
+                }
             }
-            x = 0;
-            inline while (x < 5) : (x += 1) {
-                s[x + y * 5] = c[x] ^ (~c[M5[x + 1]] & c[M5[x + 2]]);
+            const full_blocks = out[0 .. out.len - out.len % State.rate];
+            if (full_blocks.len > 0) {
+                self.st.squeeze(full_blocks);
```

Comment:
If we absorb some blocks and a bit, won't `self.st.squeeze` be called twice which causes `0x80` to be absorbed twice?

Reply:
Oh, right! Good catch!

That's actually why I introduced `squeeze_more()` in addition to `squeeze()` in the Xoodyak implementation.

Gonna do the same thing here. Thanks!

Reply:
Done, thanks!